### PR TITLE
HDDS-9626. [Recon] Disk Usage page with high number of key/bucket/volume

### PR DESCRIPTION
**Pull request by smitajoshi12**
## What changes were proposed in this pull request?

When the number of keys/volume/bucket are huge, the current disk usage UI doesnt make much sense.
This pull request introduces enhancements to the Recon disk usage endpoint to significantly improve usability and performance when dealing with large datasets:
**Top Entities Focus:** The endpoint has been updated to efficiently sort and display only the top entities by size. This targeted approach helps users easily identify the most significant space consumers, addressing the impracticality of visualizing thousands of records in a single view.
Efficient Sorting with Parallel Streams: To manage and sort vast numbers of records effectively, we've implemented parallel stream processing.

**Key advantages of using parallel streams include :-**
Better Utilization of Multi-core Processors: Enables concurrent sorting operations across multiple cores, drastically cutting down processing times for large datasets.
Optimized for Large Datasets: The parallelism overhead is more efficiently distributed over a large number of elements, making it particularly suited for our use case.

**Backend PR For Reference:-**
https://github.com/apache/ozone/pull/6318

## What is the link to the Apache JIRA
https://issues.apache.org/jira/browse/HDDS-9626

## How was this patch tested?
Manually
Before this PR
image

After this PR
Tested with Cluster Data
image

image

image

### CODE REVIEW

**Comment by swamirishi** [comment]
@dombizita @devmadhuu Can you please take a look at this patch.

**Review by ArafatKhan2198 — COMMENTED** [review]
Thanks for working on this @smitajoshi12 
While testing this patch locally I noticed a few discrepancies while setting the Display Limit :- 

- I currently have **56 keys** in my cluster all of which are present inside the `buckettest`.

### 1. When I set the display limit to 5 I notice that 5 objects of the highest size is displayed and also the remaining objects are clubbed inside the `Other Objects` 

image

### 2. For 20 I get the correct result as well :- 

image

### 3. But when I set the limit to 30 I do not see the `Other Objects` slot anywhere even though there are a total of 56 keys hence the remaining 26 Keys need to get clubbed into `Other Objects`.

image

**Comment by smitajoshi12** [comment]
> Thanks for working on this @smitajoshi12 While testing this patch locally I noticed a few discrepancies while setting the Display Limit :-
> 
> * I currently have **56 keys** in my cluster all of which are present inside the `buckettest`.
> 
> ### 1. When I set the display limit to 5 I notice that 5 objects of the highest size is displayed and also the remaining objects are clubbed inside the `Other Objects`
> image
> 
> ### 2. For 20 I get the correct result as well :-
> image
> 
> ### 3. But when I set the limit to 30 I do not see the `Other Objects` slot anywhere even though there are a total of 56 keys hence the remaining 26 Keys need to get clubbed into `Other Objects`.
> image

@ArafatKhan2198 image

**Review by ArafatKhan2198 — CHANGES_REQUESTED** [review]
Thanks for updating the patch, @smitajoshi12. We are now using the correct API parameters for sorting the subpaths, but there is still an issue from the UI perspective. Let's say we have three files:

```
file1 -> Size -> 1 KB
file2 -> Size -> 10 KB
file3 -> Size -> 1 GB
```

The API endpoint would return a response in descending order of size. However, the problem is that the UI representation becomes skewed, as shown in the image below:
Here, we have three directories with sizes 1 KB, 10 KB, and 1 GB. I believe the size of each part of the pie chart is relative to the file size, but this creates a poor user experience. We need to address this issue to improve the user interface.

image

Could you please take care of this!

**Comment by smitajoshi12** [comment]
> Thanks for updating the patch, @smitajoshi12. We are now using the correct API parameters for sorting the subpaths, but there is still an issue from the UI perspective. Let's say we have three files:
> 
> ```
> file1 -> Size -> 1 KB
> file2 -> Size -> 10 KB
> file3 -> Size -> 1 GB
> ```
> 
> The API endpoint would return a response in descending order of size. However, the problem is that the UI representation becomes skewed, as shown in the image below: Here, we have three directories with sizes 1 KB, 10 KB, and 1 GB. I believe the size of each part of the pie chart is relative to the file size, but this creates a poor user experience. We need to address this issue to improve the user interface.
> 
> image
> Could you please take care of this!

@ArafatKhan2198 
Can we raise seprate JIRA  for it as it is known issue  need to work on lots of changes.

**Review by ArafatKhan2198 — APPROVED** [review]
Thanks, @smitajoshi12 for working on this. 
LGTM

**Review by devmadhuu — APPROVED** [review]
Thanks @smitajoshi12 for working on this. LGTM +1

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/diskUsage/diskUsage.tsx [hunk 1]
```diff
@@ -144,20 +145,19 @@ export class DiskUsage extends React.Component<Record<string, object>, IDUState>
       const dataSize = duResponse.size;
       let subpaths: IDUSubpath[] = duResponse.subPaths;
 
-      subpaths.sort((a, b) => (a.size < b.size) ? 1 : -1);
-
       // Only show top n blocks with the most DU,
       // other blocks are merged as a single block
-      if (subpaths.length > limit) {
+      if (subpaths.length > limit || (subpaths.length > 0 && limit === MAX_DISPLAY_LIMIT)) {
```
Comment (dombizita) [inline hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/diskUsage/diskUsage.tsx:150]: @smitajoshi12 Could you explain why we need the other condition? 

`(subpaths.length > 0 && limit === MAX_DISPLAY_LIMIT)` - we have any number (not zero) of subpaths and the limit is set to 30, the max display limit. This means that even even if we don't have more subpaths than the limit, we are adding the `Other Objects`? Why is that needed?

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/diskUsage/diskUsage.tsx [hunk 2]
```diff
@@ -144,20 +145,19 @@ export class DiskUsage extends React.Component<Record<string, object>, IDUState>
...
         subpaths = subpaths.slice(0, limit);
         let topSize = 0;
-        for (let i = 0; i < limit; ++i) {
+        for (let i = 0; limit === MAX_DISPLAY_LIMIT ? i < subpaths.length : i < limit; ++i) {
```
Comment (dombizita) [inline hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/diskUsage/diskUsage.tsx:153]: I think this could work wrong. What happens if the limit is set to 10, but we only have 5 subpaths? In that case we would iterate until `i` is smaller than 10? But we don't have that many elements in `subpaths`. Did I miss something?

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/diskUsage/diskUsage.tsx [hunk 3]
```diff
@@ -265,8 +265,8 @@ export class DiskUsage extends React.Component<Record<string, object>, IDUState>
 
   updateDisplayLimit(e): void {
     let res = -1;
-    if (e.key === 'all') {
-      res = Number.MAX_VALUE;
+    if (e.key === '30') {
+      res = Number.parseInt(e.key, 10);
     } else {
       res = Number.parseInt(e.key, 10);
     }
```
Comment (dombizita) [inline hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/diskUsage/diskUsage.tsx:272]: I don't understand this. With the changes in both cases (if `e.key` is `'all'` and if it's `'30'`) we will set `res = Number.parseInt(e.key, 10);`

**Review by dombizita — CHANGES_REQUESTED** [review]

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/diskUsage/diskUsage.tsx (hunk 1 above)
Comment (smitajoshi12) [reply hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/diskUsage/diskUsage.tsx:150]: @dombizita 

1)subpaths.length > 0 && limit === MAX_DISPLAY_LIMIT  this condition is purely for Limit=30 then we showing other objects in Pie Chart

2) Previous Conditions are for Example out of 30 in from Dropdown we have selected to show only 10 Enties in Pie Chart Length=30 and our limit=10 then other 20 entities we are showing as other object.

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/diskUsage/diskUsage.tsx (hunk 3 above)
Comment (smitajoshi12) [reply hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/diskUsage/diskUsage.tsx:272]: @dombizita 

In Old design earlier It was  All in dropdown as now we are setting with Max limit=30
subpaths.length > 0 && limit === MAX_DISPLAY_LIMIT

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/diskUsage/diskUsage.tsx (hunk 2 above)
Comment (smitajoshi12) [reply hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/diskUsage/diskUsage.tsx:153]: @dombizita 

If Limit=10 and subpath.length=5 then It will not enter in Outer If Condition
5 > 10 it will become false
if (subpaths.length > limit || (subpaths.length > 0 && limit === MAX_DISPLAY_LIMIT)) {
              subpaths = subpaths.slice(0, limit);
              let topSize = 0;
              **for (let i = 0; limit === MAX_DISPLAY_LIMIT ? i < subpaths.length : i < limit; ++i)** {
                topSize += subpaths[i].size;
              }
              const otherSize = dataSize - topSize;
              if (otherSize > 0) {
                const other: IDUSubpath = {path: OTHER_PATH_NAME, size: otherSize};
                subpaths.push(other);
              }
      }

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/diskUsage/diskUsage.tsx [hunk 4]
```diff
@@ -144,20 +145,19 @@ export class DiskUsage extends React.Component<Record<string, object>, IDUState>
       const dataSize = duResponse.size;
       let subpaths: IDUSubpath[] = duResponse.subPaths;
 
-      subpaths.sort((a, b) => (a.size < b.size) ? 1 : -1);
-
-      // Only show top n blocks with the most DU,
-      // other blocks are merged as a single block
-      if (subpaths.length > limit) {
+      // This logic is for other objects for Max Limit 30 we are calculating 
+      // Other Object Size = Total Size - Sum of all subpaths Size
```
Comment (dombizita) [inline hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/diskUsage/diskUsage.tsx:149]: I don't understand this comment. I'd go with something like this:
```suggestion
      // We need to calculate the size of "Other objects" in two cases: 
      // 1) If we have more subpaths listed, than the limit.  
      // 2) If the limit is set to the maximum limit (30) and we have any number of subpaths. In this case we won't
      // necessarily have "Other objects", but later we check if the other objects's size is more than zero (we will have
      // other objects if there are more than 30 subpaths, but we can't check on that, as the response will always have 
      // 30 subpaths, but from the total size and the subpaths size we can calculate it).
```

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/diskUsage/diskUsage.tsx [hunk 5]
```diff
@@ -144,20 +145,19 @@ export class DiskUsage extends React.Component<Record<string, object>, IDUState>
...
+      if (subpaths.length >= limit || (subpaths.length > 0 && limit === MAX_DISPLAY_LIMIT)) {
```
Comment (dombizita) [inline hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/diskUsage/diskUsage.tsx:150]: Why do we need the `>=` in the `subpaths.length >= limit` condition? If we have 10 subpaths and 10 is the limit, we won't need "Other objects", right?

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/diskUsage/diskUsage.tsx [hunk 6]
```diff
@@ -549,7 +546,7 @@ export class DiskUsage extends React.Component<Record<string, object>, IDUState>
                   </div>
                   <div className='dropdown-button'>
                     <Dropdown overlay={menu} placement='bottomCenter'>
-                      <Button>Display Limit: {(displayLimit === Number.MAX_VALUE) ? 'All' : displayLimit}</Button>
+                      <Button>Display Limit: {(displayLimit === Number.MAX_VALUE) ? MAX_DISPLAY_LIMIT : displayLimit}</Button>
```
Comment (dombizita) [inline hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/diskUsage/diskUsage.tsx:549]: I think this has some from the previous code. Isn't this enough? 
```suggestion
                      <Button>Display Limit: {displayLimit}</Button>
```

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/diskUsage/diskUsage.tsx [hunk 7]
```diff
@@ -265,12 +265,9 @@ export class DiskUsage extends React.Component<Record<string, object>, IDUState>
 
   updateDisplayLimit(e): void {
     let res = -1;
-    if (e.key === 'all') {
-      res = Number.MAX_VALUE;
-    } else {
+    if (e.key) {
       res = Number.parseInt(e.key, 10);
     }
-
```
Comment (dombizita) [inline hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/diskUsage/diskUsage.tsx:273]: Why do we need all of this? Can't we just set the `res` like this?
`res = Number.parseInt(e.key, 10);`

**Review by dombizita — CHANGES_REQUESTED** [review]
Thanks for updating your patch @smitajoshi12! Please take a look at my comments!

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/diskUsage/diskUsage.tsx (hunk 5 above)
Comment (smitajoshi12) [reply hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/diskUsage/diskUsage.tsx:150]: @dombizita  Other Objects are depending on Total size and sum of all subpaths Size.
Example 1 
Limit=10 and subpaths.length=10
Total Size =100KB Sum of all 10 nodes size=90KB then we need to show remaining 10 in other objects to cover this test case checking condition  >=

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/diskUsage/diskUsage.tsx (hunk 6 above)
Comment (smitajoshi12) [reply hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/diskUsage/diskUsage.tsx]: @dombizita 
Updated code in latest commit missed this comment earlier review.

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/diskUsage/diskUsage.tsx (hunk 4 above)
Comment (smitajoshi12) [reply hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/diskUsage/diskUsage.tsx]: @dombizita 
Updated comment line.

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/diskUsage/diskUsage.tsx (hunk 5 above)
Comment (dombizita) [reply hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/diskUsage/diskUsage.tsx:155]: In your example the `subpaths.length` is 10. In that case the sum of that can't be less, than the total. The subpaths's size is either the total size or if we have 30 subpaths, we have the chance that we have more subpaths, in that case we need to sum them up and subtract it from the total size. 

If the `subpaths.length` is not the maximum limit, than we won't need other objects, unless the limit is less than the number of subpaths.

Comment (dombizita) [reply hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/diskUsage/diskUsage.tsx:155]: Please correct me if I'm wrong @ArafatKhan2198 @devmadhuu

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/diskUsage/diskUsage.tsx (hunk 7 above)
Comment (smitajoshi12) [reply hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/diskUsage/diskUsage.tsx]: @dombizita 
Thanks Done changes as suggested by you.

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/diskUsage/diskUsage.tsx (hunk 5 above)
Comment (smitajoshi12) [reply hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/diskUsage/diskUsage.tsx]: @dombizita 
Done changes as subpaths.length > limit in latest commit.

File: hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/diskUsage/diskUsage.tsx (hunk 7 above)
Comment (smitajoshi12) [reply hadoop-ozone/recon/src/main/resources/webapps/recon/ozone-recon-web/src/views/diskUsage/diskUsage.tsx]: > @smitajoshi12 are all comments addressed given by Zita ?

@devmadhuu 

Yes Devesh all comments addressed . you can merge.

**Review by dombizita — APPROVED** [review]

**Comment by devmadhuu** [comment]
Thanks @smitajoshi12 for working on this patch. Thanks @dombizita , @ArafatKhan2198 for reviewing the patch.
